File: snippets/PlacesSnippets/PlacesSnippets/PlaceAutocomplete.m
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


// [START maps_ios_dev_guides_place_autocomplete_fullscreen]
#import "PlaceAutocomplete.h"
@import GooglePlaces;

@interface PlaceAutocomplete () <GMSAutocompleteViewControllerDelegate>

@end

@implementation PlaceAutocomplete

- (void)viewDidLoad {
  [super viewDidLoad];
  [self makeButton];
}

// Present the autocomplete view controller when the button is pressed.
- (void)autocompleteClicked {
  GMSAutocompleteViewController *acController = [[GMSAutocompleteViewController alloc] init];
  acController.delegate = self;

  // Specify the place data types to return.
  GMSPlaceField fields = (GMSPlaceFieldName | GMSPlaceFieldPlaceID);
  acController.placeFields = fields;

  // Specify a filter.
  GMSAutocompleteFilter *filter = [[GMSAutocompleteFilter alloc] init];
  filter.type = kGMSPlacesAutocompleteTypeFilterAddress;
  acController.autocompleteFilter = filter;

  // Display the autocomplete view controller.
  [self presentViewController:acController animated:YES completion:nil];
}

// Add a button to the view.
- (void)makeButton{
  UIButton *btnLaunchAc = [UIButton buttonWithType:UIButtonTypeCustom];
  [btnLaunchAc addTarget:self
                  action:NSSelectorFromString(@"autocompleteClicked")
        forControlEvents:UIControlEventTouchUpInside];
  [btnLaunchAc setTitle:@"Launch autocomplete" forState:UIControlStateNormal];
  btnLaunchAc.frame = CGRectMake(5.0, 150.0, 300.0, 35.0);
  btnLaunchAc.backgroundColor = [UIColor blueColor];
  [self.view addSubview:btnLaunchAc];
}

// Handle the user's selection.
- (void)viewController:(GMSAutocompleteViewController *)viewController didAutocompleteWithPlace:(GMSPlace *)place {
  [self dismissViewControllerAnimated:YES completion:nil];
  // Do something with the selected place.
  NSLog(@"Place name %@", place.name);
  NSLog(@"Place ID %@", place.placeID);
  NSLog(@"Place attributions %@", place.attributions.string);
}

- (void)viewController:(GMSAutocompleteViewController *)viewController didFailAutocompleteWithError:(NSError *)error {
  [self dismissViewControllerAnimated:YES completion:nil];
  // TODO: handle the error.
  NSLog(@"Error: %@", [error description]);
}

// User canceled the operation.
- (void)wasCancelled:(GMSAutocompleteViewController *)viewController {
  [self dismissViewControllerAnimated:YES completion:nil];
}

// Turn the network activity indicator on and off again.
- (void)didRequestAutocompletePredictions:(GMSAutocompleteViewController *)viewController {
  [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
}

- (void)didUpdateAutocompletePredictions:(GMSAutocompleteViewController *)viewController {
  [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
}

@end
// [END maps_ios_dev_guides_place_autocomplete_fullscreen]

@interface PlaceAutocomplete2 : UIViewController <GMSAutocompleteResultsViewControllerDelegate>

@end

@implementation PlaceAutocomplete2

// [START maps_ios_dev_guides_place_autocomplete_results]
GMSAutocompleteResultsViewController *_resultsViewController;
UISearchController *_searchController;

- (void)viewDidLoad {
  [super viewDidLoad];

  // ...
  _resultsViewController = [[GMSAutocompleteResultsViewController alloc] init];
  _resultsViewController.delegate = self;

  _searchController = [[UISearchController alloc]
                       initWithSearchResultsController:_resultsViewController];
  _searchController.searchResultsUpdater = _resultsViewController;

  // Put the search bar in the navigation bar.
  [_searchController.searchBar sizeToFit];
  self.navigationItem.titleView = _searchController.searchBar;

  // When UISearchController presents the results view, present it in
  // this view controller, not one further up the chain.
  self.definesPresentationContext = YES;

  // Prevent the navigation bar from being hidden when searching.
  _searchController.hidesNavigationBarDuringPresentation = NO;
}

// Handle the user's selection.
- (void)resultsController:(GMSAutocompleteResultsViewController *)resultsController didAutocompleteWithPlace:(GMSPlace *)place {
    _searchController.active = NO;
    // Do something with the selected place.
    NSLog(@"Place name %@", place.name);
    NSLog(@"Place address %@", place.formattedAddress);
    NSLog(@"Place attributions %@", place.attributions.string);
}

- (void)resultsController:(GMSAutocompleteResultsViewController *)resultsController didFailAutocompleteWithError:(NSError *)error {
  [self dismissViewControllerAnimated:YES completion:nil];
  // TODO: handle the error.
  NSLog(@"Error: %@", [error description]);
}

// Turn the network activity indicator on and off again.
- (void)didRequestAutocompletePredictionsForResultsController:(GMSAutocompleteResultsViewController *)resultsController {
  [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
}

- (void)didUpdateAutocompletePredictionsForResultsController:(GMSAutocompleteResultsViewController *)resultsController {
  [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
}
// [END maps_ios_dev_guides_place_autocomplete_results]

- (void)navBar {
  // [START maps_ios_dev_guides_place_autocomplete_nav_bar]
  self.navigationController.navigationBar.translucent = NO;
  _searchController.hidesNavigationBarDuringPresentation = NO;

  // This makes the view area include the nav bar even though it is opaque.
  // Adjust the view placement down.
  self.extendedLayoutIncludesOpaqueBars = YES;
  self.edgesForExtendedLayout = UIRectEdgeTop;
  // [END maps_ios_dev_guides_place_autocomplete_nav_bar]
}

@end

@interface PlaceAutocomplete3 : UIViewController

@end

@implementation PlaceAutocomplete3

// [START maps_ios_dev_guides_place_autocomplete_search_bar]
UISearchController *_searchController;

- (void)viewDidLoad {
  [super viewDidLoad];

  // ...

  // Add a search bar
  UIView *subView = [[UIView alloc] initWithFrame:CGRectMake(0, 65.0, 250, 50)];
  [subView addSubview:_searchController.searchBar];
  [_searchController.searchBar sizeToFit];
  [self.view addSubview:subView];
}
// [END maps_ios_dev_guides_place_autocomplete_search_bar]

@end

@interface PlaceAutocomplete4 : UIViewController

@end

@implementation PlaceAutocomplete4

UISearchController *_searchController;

// [START maps_ios_dev_guides_place_autocomplete_search_bar_popover]
- (void)viewDidLoad {
  [super viewDidLoad];

  // ...

  // Add the search bar to the right of the nav bar,
  // use a popover to display the results.
  // Set an explicit size as we don't want to use the entire nav bar.
  _searchController.searchBar.frame = CGRectMake(0, 0, 250.0f, 44.0f);
  self.navigationItem.rightBarButtonItem =
  [[UIBarButtonItem alloc] initWithCustomView:_searchController.searchBar];

  // When UISearchController presents the results view, present it in
  // this view controller, not one further up the chain.
  self.definesPresentationContext = YES;

  // Keep the navigation bar visible.
  _searchController.hidesNavigationBarDuringPresentation = NO;

  _searchController.modalPresentationStyle = UIModalPresentationPopover;
}
// [END maps_ios_dev_guides_place_autocomplete_search_bar_popover]

- (void)styling {
  // [START maps_ios_dev_guides_place_autocomplete_appearance]
  UITextField *textField = [UITextField appearanceWhenContainedInInstancesOfClasses:@[[UISearchBar class]]];
  [textField setDefaultTextAttributes:@{NSForegroundColorAttributeName:[UIColor greenColor]}];
  // [END maps_ios_dev_guides_place_autocomplete_appearance]

  // [START maps_ios_dev_guides_place_autocomplete_appearance2]
  // Define some colors.
  UIColor *darkGray = [UIColor darkGrayColor];
  UIColor *lightGray = [UIColor lightGrayColor];

  // Navigation bar background.
  [[UINavigationBar appearance] setBarTintColor:darkGray];
  [[UINavigationBar appearance] setTintColor:lightGray];

  // Color of typed text in the search bar.
  NSDictionary *searchBarTextAttributes = @{
    NSForegroundColorAttributeName: lightGray,
    NSFontAttributeName : [UIFont systemFontOfSize:[UIFont systemFontSize]]
  };
  UITextField *appearance = [UITextField appearanceWhenContainedInInstancesOfClasses:@[[UISearchBar class]]];
  appearance.defaultTextAttributes = searchBarTextAttributes;

  // Color of the placeholder text in the search bar prior to text entry.
  NSDictionary *placeholderAttributes = @{
    NSForegroundColorAttributeName: lightGray,
    NSFontAttributeName : [UIFont systemFontOfSize:[UIFont systemFontSize]]
  };

  // Color of the default search text.
  // NOTE: In a production scenario, "Search" would be a localized string.
  appearance.attributedPlaceholder = [[NSAttributedString alloc] initWithString:@"Search"
                                                                     attributes:placeholderAttributes];

  // Color of the in-progress spinner.
  [[UIActivityIndicatorView appearance] setColor:lightGray];

  // To style the two image icons in the search bar (the magnifying glass
  // icon and the 'clear text' icon), replace them with different images.
  [[UISearchBar appearance] setImage:[UIImage imageNamed:@"custom_clear_x_high"]
                    forSearchBarIcon:UISearchBarIconClear
                              state:UIControlStateHighlighted];
  [[UISearchBar appearance] setImage:[UIImage imageNamed:@"custom_clear_x"]
                    forSearchBarIcon:UISearchBarIconClear
                              state:UIControlStateNormal];
  [[UISearchBar appearance] setImage:[UIImage imageNamed:@"custom_search"]
                      forSearchBarIcon:UISearchBarIconSearch
                              state:UIControlStateNormal];

  // Color of selected table cells.
  UIView *selectedBackgroundView = [[UIView alloc] init];
  selectedBackgroundView.backgroundColor = [UIColor lightGrayColor];
  UITableViewCell *tableViewCell = [UITableViewCell appearanceWhenContainedInInstancesOfClasses:@[[GMSAutocompleteViewController class]]];
  tableViewCell.selectedBackgroundView = selectedBackgroundView;
  // [END maps_ios_dev_guides_place_autocomplete_appearance2]
}

@end

File: snippets/PlacesSnippets/PlacesSnippets/Swift/PlaceAutocomplete.swift
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


// [START maps_ios_dev_guides_place_autocomplete_fullscreen]
import GooglePlaces

class PlaceAutocomplete: UIViewController {

  override func viewDidLoad() {
    super.viewDidLoad()
    makeButton()
  }

  // Present the Autocomplete view controller when the button is pressed.
  @objc func autocompleteClicked(_ sender: UIButton) {
    let autocompleteController = GMSAutocompleteViewController()
    autocompleteController.delegate = self

    // Specify the place data types to return.
    let fields: GMSPlaceField = GMSPlaceField(
      rawValue:GMSPlaceField.name.rawValue | GMSPlaceField.placeID.rawValue
    )!
    autocompleteController.placeFields = fields

    // Specify a filter.
    let filter = GMSAutocompleteFilter()
    filter.type = .address
    autocompleteController.autocompleteFilter = filter

    // Display the autocomplete view controller.
    present(autocompleteController, animated: true, completion: nil)
  }

  // Add a button to the view.
  func makeButton() {
    let btnLaunchAc = UIButton(frame: CGRect(x: 5, y: 150, width: 300, height: 35))
    btnLaunchAc.backgroundColor = .blue
    btnLaunchAc.setTitle("Launch autocomplete", for: .normal)
    btnLaunchAc.addTarget(self, action: #selector(autocompleteClicked), for: .touchUpInside)
    self.view.addSubview(btnLaunchAc)
  }
}

extension PlaceAutocomplete: GMSAutocompleteViewControllerDelegate {

  // Handle the user's selection.
  func viewController(_ viewController: GMSAutocompleteViewController, didAutocompleteWith place: GMSPlace) {
    print("Place name: \(String(describing: place.name))")
    print("Place ID: \(place.placeID ?? "unknown")")
    print("Place attributions: \(String(describing: place.attributions))")
    dismiss(animated: true, completion: nil)
  }

  func viewController(_ viewController: GMSAutocompleteViewController, didFailAutocompleteWithError error: Error) {
    // TODO: handle the error.
    print("Error: ", error.localizedDescription)
  }

  // User canceled the operation.
  func wasCancelled(_ viewController: GMSAutocompleteViewController) {
    dismiss(animated: true, completion: nil)
  }

  // Turn the network activity indicator on and off again.
  func didRequestAutocompletePredictions(_ viewController: GMSAutocompleteViewController) {
    UIApplication.shared.isNetworkActivityIndicatorVisible = true
  }

  func didUpdateAutocompletePredictions(_ viewController: GMSAutocompleteViewController) {
    UIApplication.shared.isNetworkActivityIndicatorVisible = false
  }
}
// [END maps_ios_dev_guides_place_autocomplete_fullscreen]

class PlaceAutocomplete2: UIViewController, GMSAutocompleteResultsViewControllerDelegate {

  // [START maps_ios_dev_guides_place_autocomplete_results]
  var resultsViewController: GMSAutocompleteResultsViewController!
  var searchController: UISearchController!
  var resultView: UITextView!

  override func viewDidLoad() {
    super.viewDidLoad()

    // ...

    resultsViewController = GMSAutocompleteResultsViewController()
    resultsViewController.delegate = self

    searchController = UISearchController(searchResultsController: resultsViewController)
    searchController.searchResultsUpdater = resultsViewController

    // Put the search bar in the navigation bar.
    searchController.searchBar.sizeToFit()
    navigationItem.titleView = searchController?.searchBar

    // When UISearchController presents the results view, present it in
    // this view controller, not one further up the chain.
    definesPresentationContext = true

    // Prevent the navigation bar from being hidden when searching.
    searchController.hidesNavigationBarDuringPresentation = false
  }

  func resultsController(_ resultsController: GMSAutocompleteResultsViewController, didAutocompleteWith place: GMSPlace) {
    searchController.isActive = false
    // Do something with the selected place.
    print("Place name: \(String(describing: place.name))")
    print("Place address: \(String(describing: place.formattedAddress))")
    print("Place attributions: \(String(describing: place.attributions))")
  }

  func resultsController(_ resultsController: GMSAutocompleteResultsViewController, didFailAutocompleteWithError error: Error){
    // TODO: handle the error.
    print("Error: ", error.localizedDescription)
  }

  // Turn the network activity indicator on and off again.
  func didRequestAutocompletePredictions(_ viewController: GMSAutocompleteViewController) {
    UIApplication.shared.isNetworkActivityIndicatorVisible = true
  }

  func didUpdateAutocompletePredictions(_ viewController: GMSAutocompleteViewController) {
    UIApplication.shared.isNetworkActivityIndicatorVisible = false
  }
  // [END maps_ios_dev_guides_place_autocomplete_results]

  func navBar() {
    // [START maps_ios_dev_guides_place_autocomplete_nav_bar]
    navigationController?.navigationBar.isTranslucent = false
    searchController.hidesNavigationBarDuringPresentation = false

    // This makes the view area include the nav bar even though it is opaque.
    // Adjust the view placement down.
    self.extendedLayoutIncludesOpaqueBars = true
    self.edgesForExtendedLayout = .top
    // [END maps_ios_dev_guides_place_autocomplete_nav_bar]
  }
}

class PlaceAutocomplete3: UIViewController {
  // [START maps_ios_dev_guides_place_autocomplete_search_bar]
  var searchController: UISearchController!

  override func viewDidLoad() {
    super.viewDidLoad()

    // ...

    let subView = UIView(frame: CGRect(x: 0, y: 65.0, width: 350.0, height: 45.0))

    subView.addSubview((searchController?.searchBar)!)
    view.addSubview(subView)
    searchController?.searchBar.sizeToFit()
  }
  // [END maps_ios_dev_guides_place_autocomplete_search_bar]
}

class PlaceAutocomplete4: UIViewController {

  var searchController: UISearchController!

  // [START maps_ios_dev_guides_place_autocomplete_search_bar_popover]
  override func viewDidLoad() {
    super.viewDidLoad()

    // ...

    // Add the search bar to the right of the nav bar,
    // use a popover to display the results.
    // Set an explicit size as we don't want to use the entire nav bar.
    searchController.searchBar.frame = (CGRect(x: 0, y: 0, width: 250.0, height: 44.0))
    navigationItem.rightBarButtonItem = UIBarButtonItem(customView: (searchController?.searchBar)!)

    // When UISearchController presents the results view, present it in
    // this view controller, not one further up the chain.
    definesPresentationContext = true

    // Keep the navigation bar visible.
    searchController.hidesNavigationBarDuringPresentation = false
    searchController.modalPresentationStyle = .popover
  }
  // [END maps_ios_dev_guides_place_autocomplete_search_bar_popover]

  func styling() {
    // [START maps_ios_dev_guides_place_autocomplete_appearance]
    let textField = UITextField.appearance(whenContainedInInstancesOf: [UISearchBar.self])
    textField.defaultTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.green]
    // [END maps_ios_dev_guides_place_autocomplete_appearance]

    // [START maps_ios_dev_guides_place_autocomplete_appearance2]
    // Define some colors.
    let darkGray = UIColor.darkGray
    let lightGray = UIColor.lightGray

    // Navigation bar background.
    let navBarAppearance = UINavigationBar.appearance()
    navBarAppearance.barTintColor = darkGray
    navBarAppearance.tintColor = lightGray

    // Color of typed text in the search bar.
    let searchBarTextAttributes = [
      NSAttributedString.Key.foregroundColor : lightGray,
      NSAttributedString.Key.font : UIFont.systemFont(ofSize: UIFont.systemFontSize)
    ]
    let appearance = UITextField.appearance(whenContainedInInstancesOf: [UISearchBar.self])
    appearance.defaultTextAttributes = searchBarTextAttributes

    // Color of the placeholder text in the search bar prior to text entry.
    let placeholderAttributes = [
      NSAttributedString.Key.foregroundColor: lightGray,
      NSAttributedString.Key.font: UIFont.systemFont(ofSize: UIFont.systemFontSize)
    ]

    // Color of the default search text.
    // NOTE: In a production scenario, "Search" would be a localized string.
    appearance.attributedPlaceholder = NSAttributedString(string: "Search", attributes: placeholderAttributes)

    // Color of the in-progress spinner.
    UIActivityIndicatorView.appearance().color = lightGray

    // To style the two image icons in the search bar (the magnifying glass
    // icon and the 'clear text' icon), replace them with different images.
    UISearchBar.appearance().setImage(
      UIImage(named: "custom_clear_x_high"), for: .clear, state: .highlighted
    )
    UISearchBar.appearance().setImage(
      UIImage(named: "custom_clear_x"), for: .clear, state: .normal
    )
    UISearchBar.appearance().setImage(
      UIImage(named: "custom_search"), for: .search, state: .normal
    )

    // Color of selected table cells.
    let selectedBackgroundView = UIView()
    selectedBackgroundView.backgroundColor = lightGray
    let tableViewCell = UITableViewCell.appearance(whenContainedInInstancesOf: [GMSAutocompleteViewController.self])
    tableViewCell.selectedBackgroundView = selectedBackgroundView
    // [END maps_ios_dev_guides_place_autocomplete_appearance2]
  }
}
